packages/tpetra/core/src/Tpetra_FECrsMatrix_decl.hpp

```diff
@@ -0,0 +1,261 @@
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+#ifndef TPETRA_FECRSMATRIX_DECL_HPP
+#define TPETRA_FECRSMATRIX_DECL_HPP
+
+
+/// \file Tpetra_FECrsMatrix_decl.hpp
+/// \brief Declaration of the Tpetra::FECrsMatrix class
+///
+/// If you want to use Tpetra::FECrsMatrix, include
+/// "Tpetra_FECrsMatrix.hpp" (a file which CMake generates and installs
+/// for you).  If you only want the declaration of Tpetra::FECrsMatrix,
+/// include this file (Tpetra_FECrsMatrix_decl.hpp).
+
+#include "Tpetra_CrsMatrix_decl.hpp"
+#include "Tpetra_FECrsGraph.hpp"
+
+namespace Tpetra {
+
+
+
+// \class FECrsMatrix
+// \brief Sparse matrix that presents a row-oriented interface that lets
+//        users read or modify entries.
+template<class Scalar        = ::Tpetra::Details::DefaultTypes::scalar_type,
+         class LocalOrdinal  = ::Tpetra::Details::DefaultTypes::local_ordinal_type,
+         class GlobalOrdinal = ::Tpetra::Details::DefaultTypes::global_ordinal_type,
+         class Node          = ::Tpetra::Details::DefaultTypes::node_type>
+class FECrsMatrix :
+    public CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>
+{
+  friend class CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
+public:
+    //! @name Typedefs
+    //@{
+
+    /// \brief This class' first template parameter; the type of each
+    ///        entry in the matrix.
+    typedef Scalar scalar_type;
```

Comment:
It would be nicer to pull these typedefs from the base class CrsMatrix.  That way, when we finally purge the last three template parameters, we won't have to change this code.

Reply:
Done